### PR TITLE
Add r_core_cmd_tobuf() and overhaul alias system

### PR DESCRIPTION
**Pull request by Lazula**
**Checklist**

- [X] Mark this if you consider it ready to merge
- [X] I've added tests (optional)

**Description**

The alias system up to this point has been pretty bare-bones and stored as string arrays, leading to slow lookups and incorrect behavior with commands that use raw bytes like `pr`. Aliases are now stored in a hashtable mapping a string key to a structure with a data buffer, size, string flag, and command flag. Aliases can now store both strings and raw bytes. To support this, API functions have been added for running commands and getting the raw output, and escaping unprintable characters in a buffer. The alias API is also greatly expanded, making it easier to work with in the future.

### CODE REVIEW

**Comment by Lazula** [comment]
Updated `r_core_cmd_tobuf()` with new `core->cons->context` members

File: libr/core/cmd_api.c [hunk 1]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
 	}
-	return cmd->aliases.keys;
+
+	ht_pp_foreach (cmd->aliases, get_keys, keylist);
+
+	return keylist;
+}
+
+R_API void r_cmd_alias_free(RCmd *cmd) {
+	ht_pp_free(cmd->aliases);
```
Comment (trufae) [inline libr/core/cmd_api.c:206]:
```suggestion
	ht_pp_free (cmd->aliases);
```

Comment (trufae) [reply libr/core/cmd_api.c:206]:
```suggestion
	ht_pp_free (cmd->aliases);
```

File: libr/core/cmd_api.c [hunk 2]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
-		if (s) {
-			tofree = s;
-			v = (const char *)s;
-		}
+R_API RCmdAliasVal *r_cmd_alias_get(RCmd *cmd, const char *k) {
+	r_return_val_if_fail (cmd && cmd->aliases && k, NULL);
+
+	bool found;
+	RCmdAliasVal *v = ht_pp_find(cmd->aliases, k, &found);
+#if DEBUG_ALIAS
```
Comment (trufae) [inline libr/core/cmd_api.c:327]: Maybe we can get rid of all this debugging code?

File: libr/core/cmd_api.c [hunk 3]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
+	ut8* out;
+	int out_sz;
+
+	/* If appending to a string, always overwrite the trailing \0 */
+	int bytes_from_first = first->is_str
+		? first->sz - 1
+		: first->sz;
+
+	out_sz = bytes_from_first + second_sz;
+	out = malloc (out_sz);
```
Comment (trufae) [inline libr/core/cmd_api.c:349]: Check for null

File: libr/core/cmd_api.c [hunk 4]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
+	if (v_old) {
+		if (!v_old->is_data) {
+#if DEBUG_ALIAS
+			eprintf ("Won't append to existing command alias.\n");
+#endif
 			return 1;
 		}
+
+		int new_len;
+		ut8* new = alias_append_internal (&new_len, v_old, (ut8 *)a, strlen (a) + 1);
```
Comment (trufae) [inline libr/core/cmd_api.c:373]: Check for null

File: libr/core/cmd_api.c [hunk 5]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
-			matches = !strcmp (k, cmd->aliases.keys[i]);
-		}
-		if (matches) {
-			return cmd->aliases.values[i];
-		}
+/* Returns a new copy of v->data. If !v->is_str, hex escaped */
+R_API char *r_cmd_alias_val_strdup(RCmdAliasVal *v) {
+	if (v->is_str) {
+		return strdup ((char *)v->data);
+	} else {
```
Comment (trufae) [inline libr/core/cmd_api.c:422]: Unnecessary else

File: libr/core/cmd_api.c [hunk 6]
```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
+		return r_str_escape_raw (v->data, v->sz);
+	}
+}
+
+/* Returns a new copy of v->data. If !v->is_str, b64 encoded. */
+R_API char *r_cmd_alias_val_strdup_b64(RCmdAliasVal *v) {
+	if (v->is_str) {
+		return strdup ((char *)v->data);
+	} else {
```
Comment (trufae) [inline libr/core/cmd_api.c:431]: Unnecessary else

File: libr/core/cmd_debug.c [hunk 7]
```diff
@@ -4725,11 +4725,25 @@ static int cmd_debug_step (RCore *core, const char *input) {
...
 }
 
-static ut8*getFileData(RCore *core, const char *arg) {
+static ut8 *getFileData(RCore *core, const char *arg, int *sz) {
+	ut8 *out;
+	int size;
 	if (*arg == '$') {
-		return (ut8*) r_cmd_alias_get (core->rcmd, arg, 1);
+		RCmdAliasVal *v  = r_cmd_alias_get (core->rcmd, arg);
+		out = malloc (v->sz);
```
Comment (trufae) [inline libr/core/cmd_debug.c:4733]: Check for null

File: libr/core/cmd_write.c [hunk 8]
```diff
@@ -588,10 +588,13 @@ static bool cmd_wff(RCore *core, const char *input) {
 	}
 	
 	if (*a == '$') {
-		const char *res = r_cmd_alias_get (core->rcmd, a, 1);
-		if (res) {
-			buf = (ut8*)strdup (res);
-			size = strlen (res);
+		RCmdAliasVal *v = r_cmd_alias_get (core->rcmd, a+1);
+		if (v) {
+			buf = malloc (v->sz);
```
Comment (trufae) [inline libr/core/cmd_write.c:593]: Nullchk

File: libr/core/cmd_write.c [hunk 9]
```diff
@@ -1546,27 +1549,55 @@ static int wt_handler_old(void *data, const char *input) {
...
 		}
-		tmp = *str? strchr (str + 1, ' ') : NULL;
+		size_sep = *str? strchr (str + 1, ' ') : NULL;
 		if (!filename || !*filename) {
 			const char* prefix = r_config_get (core->config, "cfg.prefixdump");
 			snprintf (_fn, sizeof (_fn), "%s.0x%08"PFMT64x, prefix, poff);
 			filename = _fn;
 		}
-		if (tmp) {
+#if 0
```
Comment (trufae) [inline libr/core/cmd_write.c:1558]: Remove debug

File: libr/util/str.c [hunk 10]
```diff
@@ -1381,6 +1381,23 @@ static char *r_str_escape_(const char *buf, int dot_nl, bool parse_esc_seq, bool
...
+R_API char *r_str_escape_raw(const ut8 *buf, int sz) {
+	r_return_val_if_fail (buf, NULL);
+
+	/* Worst case scenario, we convert every byte to a \xhh escape */
+	char *new_buf = malloc (1 + sz * 4);
+	if (!new_buf) {
+		return NULL;
+	}
+	char *q = new_buf;
+	for (int i = 0; i < sz; i++) {
```
Comment (trufae) [inline libr/util/str.c:1394]: Move this int outside the for

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/core/cmd_api.c (hunk 2 above)
Comment (Lazula) [reply libr/core/cmd_api.c:327]: That's fine with me

File: libr/core/cmd_api.c (hunk 5 above)
Comment (Lazula) [reply libr/core/cmd_api.c:422]: Will remove, though imo explicit else is fine in cases like this

File: libr/core/cmd_debug.c (hunk 7 above)
Comment (Lazula) [reply libr/core/cmd_debug.c:4733]: also missed doing arg+1 here in alias_get

Comment (Lazula) [reply libr/core/cmd_debug.c:4733]: (aliases arent stored with $ anymore)

File: libr/util/str.c (hunk 10 above)
Comment (Lazula) [reply libr/util/str.c:1394]: What's wrong with loop scope? Isn't it more confusing to declare outside?

Comment (trufae) [reply libr/util/str.c:1394]: It wont compile on macos-ppc and other systems with toolchains before y2k. Its a little change that doesnt hurt much and actually it looks to me like a c++ism, () shouldnt be a scope to define variables imho

**Comment by trufae** [comment]
Ill have to take a look at the ci. Because there was a check to detect “for (int” and fail if found

File: libr/util/str.c (hunk 10 above)
Comment (Lazula) [reply libr/util/str.c:1394]: It's C99, but imo being able to access a loop variable outside is not good and leads to people doing weird things with it, and smaller scope = less clutter when things do get more complicated.

If we need to do this for C89 or other weird legacy compatibility we should note that in developers/contributing docs

**Comment by Lazula** [comment]
> Ill have to take a look at the ci. Because there was a check to detect “for (int” and fail if found

Seems strange, afaik any systems that dont support loop scope also don't support declarations not at top of block, and we do lots of that

File: libr/core/cmd_api.c [hunk 11]
```diff
@@ -185,114 +184,200 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
-				}
-				memmove (cmd->aliases.values,
-					cmd->aliases.values + 1,
-					cmd->aliases.count * sizeof (void*));
-				memmove (cmd->aliases.keys,
-					cmd->aliases.keys + 1,
-					cmd->aliases.count * sizeof (void*));
+R_API int r_cmd_alias_set_raw(RCmd *cmd, const char *k, const ut8 *v, int sz) {
+	RCmdAliasVal *val = R_NEW (RCmdAliasVal);
+	val->data = malloc (sz);
```
Comment (trufae) [inline libr/core/cmd_api.c:244]: chekc for sz > 0 and if it returns != NULL

File: libr/core/cmd_api.c [hunk 12]
```diff
@@ -185,114 +184,200 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
...
+	memcpy (val->data, v, sz);
+	val->sz = sz;
+
+	/* If it's a string already, we speed things up later by checking now */
+	const ut8 *firstnull = NULL;
+	for (int i = 0; i < sz; i++) {
```
Comment (trufae) [inline libr/core/cmd_api.c:250]: move thiss int outside the for()

File: test/db/cmd/cmd_alias [hunk 13]
```diff
@@ -57,3 +57,77 @@ EXPECT=<<EOF
...
+CMDS=<<EOF
+wx 4141414100
+ps >$str_redirect_alias
+$str_set_alias=$BBBB
+wx 430043004300
+pr 6 >$bytes_redirect_alias
+$bytes_set_alias=$D\\x00D\\x00D\\x00
+$bytes_decode_alias=base64:RQBFAEUA
+
+$*
```
Comment (trufae) [inline test/db/cmd/cmd_alias:86]: also test 'cat $alias'

**Review by trufae — CHANGES_REQUESTED** [review]

**Review by trufae — APPROVED** [review]

**Pull request merged**